src/metabase/lib/remove_replace.cljc

```diff
@@ -563,31 +563,45 @@
          (lib.equality/matching-column-sets? query stage-number fields
                                              (lib.metadata.calculation/default-columns-for-stage query stage-number)))))
 
-(defn- normalize-fields-for-join [query stage-number join]
-  (if (#{:none :all} (:fields join))
+(defn- normalize-fields-for-join [removed-location query stage-number join]
+  (cond
     ;; Nothing to do if it's already a keyword.
-    join
-    (cond-> join
-      (lib.equality/matching-column-sets?
-        query stage-number (:fields join)
-        (lib.metadata.calculation/returned-columns query stage-number (assoc join :fields :all)))
-      (assoc :fields :all))))
-
-(defn- normalize-fields-for-stage [query stage-number]
+    (#{:none :all} (:fields join)) join
+
+    ;; If it's missing, treat it as `:all` unless we just removed a field.
+    ;; TODO: This really should be a different function called by `remove-field`; it also needs to filter on the stage.
+    (and (or (= removed-location [:aggregation])
+             (= removed-location [:breakout]))
+         (not (contains? join :fields)))
+    (assoc join :fields :all)
+
+    (lib.equality/matching-column-sets?
+      query stage-number (:fields join)
+      (lib.metadata.calculation/returned-columns query stage-number (assoc join :fields :all)))
+    (assoc join :fields :all)
+
+    :else join))
+
+(defn- normalize-fields-for-stage [removed-location query stage-number]
```

Comment:
I would put `removed-location` after stage-number since so many functions follow `query, stage-number, &  args` that this is surprising
[alex miller says](https://ask.clojure.org/index.php/8373/when-should-prefer-comp-and-partial-to-anonymous-functions?show=8378#a8378) anonymous fns are more idiomatic than  `partial`